packages/core/src/scp-cf/destination-accessor.ts

```diff
@@ -482,19 +285,346 @@ const emptyDestinationByType: DestinationsByType = {
+    if (!(await this.isProviderNeeded())) {
+      return emptyDestinationByType;
+    }
+
+    if (!this.providerDestinationCache) {
+      const destinationServiceCreds = getDestinationServiceCredentials();
+
+      this.providerDestinationCache = await getInstanceAndSubaccountDestinations(
+        destinationServiceCreds.uri,
+        this.providerToken,
+        this.options
+      );
+      this.updateCacheService(
+        this.decodedProviderToken,
+        this.providerDestinationCache
+      );
+    }
+
+    return this.providerDestinationCache;
+  }
+
+  private async getSubscriberDestination() {
+    if (!this.isSubscriberNeeded()) {
+      return emptyDestinationByType;
+    }
+
+    if (!this.subscriberDestinationCache) {
+      const destinationServiceCreds = getDestinationServiceCredentials();
+
+      const accessToken = await serviceToken('destination', {
+        userJwt: this.decodedUserJwt,
+        ...this.options
+      });
+      this.subscriberDestinationCache = await getInstanceAndSubaccountDestinations(
+        destinationServiceCreds.uri,
+        accessToken,
+        this.options
+      );
+      this.updateCacheService(
+        this.decodedUserJwt!,
+        this.subscriberDestinationCache
+      );
+    }
+
+    return this.subscriberDestinationCache;
+  }
+
+  private trySubscriberCache(): Destination | undefined {
+    if (this.useCache && this.isSubscriberNeeded()) {
+      const subscriberDestinationCache = destinationCache.retrieveDestinationFromCache(
+        this.decodedUserJwt!,
+        this.name,
+        this.isolationStrategy
+      );
+      if (subscriberDestinationCache) {
+        logger.info(
+          'Successfully retrieved destination from destination service cache for subscriber destinations.'
+        );
+        return subscriberDestinationCache;
+      }
+    }
+  }
+  private isSubscriberNeeded(): boolean {
+    if (!this.decodedUserJwt) {
+      return false;
+    }
+    if (this.selectionStrategy === alwaysProvider) {
+      return false;
+    }
+    if (isIdenticalTenant(this.decodedUserJwt, this.decodedProviderToken)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private async isProviderNeeded(): Promise<boolean> {
+    if (this.selectionStrategy === alwaysSubscriber) {
+      return false;
+    }
+    if (this.selectionStrategy === alwaysProvider) {
+      return true;
+    }
+    if (
+      this.decodedUserJwt &&
+      isIdenticalTenant(this.decodedUserJwt, this.decodedProviderToken)
+    ) {
+      return true;
+    }
+    // We are in selection strategy subscriberFirst here. So if we get something from subscriber ignore provider.
+    if (await this.destinationFoundFromSubscriber()) {
+      return false;
+    }
+    return true;
+  }
+
+  private async tryProviderCache(): Promise<Destination | undefined> {
+    if (this.useCache) {
+      const providerDestinationCache = destinationCache.retrieveDestinationFromCache(
```

Comment:
I'd rename this to something that indicates that the thing is a destination, e.g. `destinationFromProviderCache` or just `destination` (since the provider-cache-context is already given by the function name).

Reply:
yes